### PR TITLE
Use annotate-snippets for emitting errors

### LINKED ISSUE #2729: Use a library for emitting errors
Preferably one shared with rustc. https://github.com/zbraniecki/annotate-snippets-rs might be such a library.

### PR DESCRIPTION
**Pull request by bash**
Resolves #2729.

There are a couple of things that I'm not sure if I did them correctly:

- `annotate-snippets` doesn't have an "internal error" annotation type. I'm currently printing "internal" as the annotation id (where the error id usually is in `rustc`):
    Screenshot 2019-04-13 at 21 49 49
- Does `rustfmt` have any guarantees about the public API when used as a library? Because I removed `fancy_print` and the `Display` impl from `FormatReport` both of which were `pub`.
- Is there a particular reason that "warning" was previously printed in red? It is now printed in yellow:
   Screenshot 2019-04-13 at 21 54 41

### CODE REVIEW

File: src/format_report_formatter.rs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        Ok(())
+    }
+}
+
+fn formatting_failure_snippet(warning_count: usize) -> Snippet {
+    Snippet {
+        title: Some(Annotation {
+            id: None,
+            label: Some(format!(
+                "rustfmt may have failed to format. See previous {} errors.",
```
Comment (scampi) [inline src/format_report_formatter.rs]: :o)

```suggestion
                "rustfmt has failed to format. See previous {} errors.",
```

**Comment by topecongiro** [comment]
@scampi 
There is an RFC which defines rustfmt stability (cc https://github.com/rust-lang/rust/issues/54504). In the RFC, [API braeking change](https://github.com/rust-lang/rfcs/blob/master/text/2437-rustfmt-stability.md#api-breaking-change) is defined as follows:

> A change that could cause a dependent crate not to build, could break a script using the executable, or breaks specification-level formatting compatibility. A formatting change in this category would be frustrating even for occasional users.

and this includes changes to the public API to rustfmt as a library. Hence, removing fancy_print and the Display impl from FormatReport is a breaking change and requires a major version update.

**Comment by topecongiro** [comment]
@bash This looks great, thank you! I really like the change you brought with this PR, so using them internally is perfectly fine. But for the reason I noted in the above comment, could you put back the removed `fancy_print` and `Display` impl for `FormatReport`?

**Comment by scampi** [comment]
@topecongiro thanks for pointing this out

File: src/format_report_formatter.rs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+impl<'a> Display for FormatReportFormatter<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let formatter = DisplayListFormatter::new(self.enable_colors);
+        let errors_by_file = &self.report.internal.borrow().0;
+
+        for (file, errors) in errors_by_file {
+            for error in errors {
+                let snippet = formatting_error_to_snippet(file, error);
+                writeln!(f, "{}", formatter.format(&DisplayList::from(snippet)))?;
```
[the remark this answers is not in the document]
Comment (bash) [reply src/format_report_formatter.rs]: Addressed in 88f1f74

**Comment by bash** [comment]
@scampi
Thanks for the review! I addressed both of your code comments.
I still need to improve/add comments to the new functions added in this PR.

**Comment by bash** [comment]
@topecongiro 
I re-implemented `fancy_print` and the `Display` impl for `FormatReport` using the new `FormatReportFormatter`. I've marked both as deprecated, is that a viable solution?

**Comment by topecongiro** [comment]
Awesome, thank you!

**Pull request merged**